core/src/main/java/hudson/node_monitors/NodesDetailsUsageStatisticsContributor.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * The MIT License
+ * 
+ * Copyright (c) 2016, Kohsuke Kawaguchi, Baptiste Mathus
+ * 
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.node_monitors;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import hudson.Extension;
+import hudson.model.Computer;
+import hudson.model.UsageStatisticsContributor;
+import hudson.node_monitors.ArchitectureMonitor.DescriptorImpl;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+
+/**
+ * Publishes informations about the used VM, number of executors, OS and so on
+ * for each node of the Jenkins cluster to the usage statistics subsystem.
+ *
+ * @see ArchitectureMonitor
+ */
+@Extension
+public class NodesDetailsUsageStatisticsContributor extends UsageStatisticsContributor {
+
+    @Override
+    public JSONObject getStatData() {
+        Jenkins j = Jenkins.getActiveInstance();
+        JSONObject o = new JSONObject();
+        List<JSONObject> nodes = new ArrayList<JSONObject>();
+        for (Computer c : j.getComputers()) {
+            JSONObject n = new JSONObject();
+            if (c.getNode() == j) {
+                n.put("master", true);
+                n.put("jvm-vendor", System.getProperty("java.vm.vendor"));
+                n.put("jvm-name", System.getProperty("java.vm.name"));
+                n.put("jvm-version", System.getProperty("java.version"));
+            }
+            n.put("executors", c.getNumExecutors());
+            DescriptorImpl descriptor = j.getDescriptorByType(DescriptorImpl.class);
+            n.put("os", descriptor.get(c));
+            nodes.add(n);
+        }
+        o.put("nodes", nodes);
```

Comment:
This means that unless you have node monitors installed, you will get no node information at all. Is that really what you want?

An alternative would be to change the extension point definition, so that the caller first constructs an empty `JSONObject`, then asks extensions to insert data into it. Thus you would not be limited to adding top-level keys. Obviously plugins need to coöperate in this case and be written defensively, but any implementation of this extension point needs to be written with care anyway to avoid sending private data.